cachi2/core/package_managers/yarn/main.py

```diff
@@ -0,0 +1,84 @@
+import logging
+
+from cachi2.core.models.input import Request
+from cachi2.core.models.output import Component, EnvironmentVariable, RequestOutput
+from cachi2.core.package_managers.yarn.project import Project, YarnRc
+from cachi2.core.package_managers.yarn.resolver import (
+    create_component_from_package,
+    resolve_dependency_info,
+)
+from cachi2.core.package_managers.yarn.utils import run_yarn_cmd
+from cachi2.core.rooted_path import RootedPath
+
+log = logging.getLogger(__name__)
+
+
+def fetch_yarn_source(request: Request) -> RequestOutput:
+    """Process all the yarn source directories in a request."""
+    components = []
+
+    for package in request.yarn_packages:
+        path = request.source_dir.join_within_root(package.path)
+        project = Project.from_source_dir(path)
+        components.extend(_resolve_yarn_project(project, request.output_dir))
+
+    return RequestOutput.from_obj_list(
+        components, _generate_environment_variables(), project_files=[]
+    )
+
+
+def _resolve_yarn_project(project: Project, output_dir: RootedPath) -> list[Component]:
+    """Process a request for a single yarn source directory.
+
+    :param project: the directory to be processed.
+    :param output_dir: the directory where the prefetched dependencies will be placed.
+    """
+    _configure_yarn_version(project)
+
+    try:
+        _prepare_for_prefetch(project.yarn_rc, output_dir)
+        packages = resolve_dependency_info(project.source_dir)
+        _fetch_dependencies(project)
+    except Exception:
+        _restore_files(project)
+        raise
+
+    _restore_files(project)
+
+    return [create_component_from_package(package, project.yarn_rc) for package in packages]
+
+
+def _configure_yarn_version(project: Project) -> None:
+    """Resolve the yarn version and set it in the package.json file if needed.
+
+    :raises PackageRejected: in case the yarn version can't be determined, or if
+    there is a mismatch between the version in package.json and yarnrc.yml.
+    """
+    pass
+
+
+def _prepare_for_prefetch(yarnrc: YarnRc, output_dir: RootedPath) -> None:
+    """Change the configuration in yarnrc to perform the prefetch.
```

Comment:
The Yarn docs [state](https://v3.yarnpkg.com/configuration/yarnrc) that "...note that most settings can also be defined through environment variables (at least for the simpler ones; arrays and objects aren't supported yet)."

So I'd prefer to only set environment variables, but my fear is that we will end up with a mixed approach, since plugins is an array and can't be set via env vars. If we do all the configuration by changing yarnrc though, it should be enough.

Wdyt? We can probably start with "plugins in yarnrc, everthing else in env vars", and change it if more things need to be changed in yarnrc.

Reply:
That does sound good, though I wonder how far down you will have to pass the env vars dict before you can pass it to `run_yarn_cmd`

Reply:
Maybe a common place (like a function?) for handling env vars would be good, otherwise they're liable to become a mess (they'll need to be used in `resolver.resolve_dependency_info`, `main._fetch_dependencies` and `main._check_yarn_cache` at least)

Reply:
> Do we need to use the env vars in resolve_dependency_info, though?
> 
> IIUC, the only configuration that would affect it is enableGlobalCache, and we won't be setting it. Is there any other config that affects yarn info?

* globalFolder - you need `yarn info` to tell you what the cachePath is
* enableMirror - same
* ignorePath
* and, to keep things sane, everything you're going to set for `yarn install`

Reply:
globalFolder will only change the cachePath if we also set enableGlobalCache, and I think we won't, right?

I do agree on the "keep things sane" bit anyways.

Reply:
You're right that globalFolder and enableMirror probably don't matter unless you also set enableGobalCache, but I don't know if we'll be setting that or not